### PR TITLE
VMware: new module: vmware_datastore_upload

### LINKED ISSUE #27893: vsphere_copy - Speed operation by writing files directly to datastore host
_From @chrrrles on August 31, 2015 16:25_

##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
vpshere_copy

##### ANSIBLE VERSION
v2.2

##### SUMMARY
If possible, we should be using pyvmomi to find the host which the datastore is connected to, and then use that with VCenter's SessionManager to generate a one time credential for a direct upload.

http://www.virtuallyghetto.com/2014/06/how-to-efficiently-transfer-files-to-datastore-in-vcenter-using-the-vsphere-api.html

http://pubs.vmware.com/vsphere-55/index.jsp#com.vmware.wssdk.apiref.doc/vim.SessionManager.html#acquireGenericServiceTicket

_Copied from original issue: ansible/ansible-modules-extras#899_

### PR DESCRIPTION
**Pull request by jeking3**
##### SUMMARY

The existing `vsphere_copy` does not directly copy to an available ESXi host (which would speed up the copy), and the copy of large files cannot be debounced.  In addition, `vsphere_copy` is the final remaining `vsphere_` module as all the rest have moved to `vmware_`.

The new module `vmware_datastore_upload` can:

1. Connect to vSphere, then connect directly to ESXi Hosts that have the datastore mounted.
    - Connecting to ESXi directly for the upload is faster.
    - If an ESXi Host is unreachable, it will try another.
1. Provide an option to debounce large file uploads.
    - This is done by leaving a `.sha256` file alongside the uploaded file and checking it later.
    - For large and unchanging files like ISOs, this can really speed things up while keeping playbooks idempotent.

This fixes #27893

##### STATUS

This is a work in progress, I wanted to get opinions on how to improve it because it is my first pull request.  There is a list of TODOs in the source that I still need to look into.  I tried to use `module_utils/urls.py` but I did not find a good mechanism to issue a PUT without wrapping `open_url` with a lot of code, so I decided to go right to `requests` instead.

For example some of the internals here could probably be moved into `module_utils/vmware.py`.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vmware_datastore_upload

##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (better-vsphere-copy 01a2a7bbe9) last updated 2018/10/18 17:07:45 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 2.7.15 (default, Sep 12 2018, 02:38:23) [GCC 6.4.0]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: s/2.x/2.8/

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Add github handle as well

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py:102]: Add a sample dict returned by module.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Are you using something specific from requests module. If yes, then try like this - https://github.com/ansible/ansible/blob/e758f69ce349c93109473c43ab0152df3b9e2120/lib/ansible/plugins/inventory/vmware_vm_inventory.py#L239

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Make `verify` configurable via parameter.

**Review by Akasurde — CHANGES_REQUESTED** [review]
Unit tests for this change would be good. Rest LGTM.

**Comment by Akasurde** [comment]
Great addition to VMware module list. cc @pdellaert @dericcrago @Im0 @ckotte @tchernomax Could you please review this ? Thanks in advance.

**Comment by jeking3** [comment]
Per the build error: `requests import found, should use ansible.module_utils.urls instead` - I looked into using `urls.py` but that is a bit of a mess, and there's no "easy" to use wrapper for issuing a PUT.  I can switch to it, but I'll have to add a put_url method in there.  I also think some of the private methods in this module can move back to `module_utils/vmware.py`, for example acquiring a service ticket.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (jeking3) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: I was following the pattern found in the google modules.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (jeking3) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: There's more to it than just that - one needs the ability to specify the client certificate, key, and CA if one is going to do certificate verification.

**Comment by jeking3** [comment]
I changed cert verification to use the module's inherited parameter 'validate_certs'.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Use https://github.com/ansible/ansible/blob/cd988f645aaf774c55a98a90e3ef42cc5b1a1563/lib/ansible/module_utils/urls.py#L1147 instead of requests.get

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Same here

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: https://github.com/ansible/ansible/blob/cd988f645aaf774c55a98a90e3ef42cc5b1a1563/lib/ansible/module_utils/urls.py#L1147 gives all options.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (ckotte) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py:44]: Could you add `aliases: [ 'datacenter_name' ]`? Lots of other modules use datacenter_name and don't have an alias defined..

**Comment by ckotte** [comment]
There's an issue with TLS:
```
TASK [esxi : Upload, overwriting any existing file at the destination] ***********************************************************************************************************************************************************************
task path: /root/ansible-vmware/playbooks/roles/esxi/tasks/main.yml:26
Trying secret FileVaultSecret(filename='/root/ansible-vmware/vault-password') for vault_id=default
Using module file /root/ansible-vmware/library/vmware-other-PRs/vmware_datastore_upload.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c '/usr/bin/python3.6 && sleep 0'
fatal: [esxi01 -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "datacenter": null,
            "datastore": "nfs_datastore",
            "debounce": "none",
            "destination": "VMware-Log-Insight-Agent-4.6.1-8587550.noarch.rpm",
            "hostname": "vcenter",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "source": "../files/VMware-Log-Insight-Agent-4.6.1-8587550.noarch.rpm",
            "username": "administrator@vsphere.local",
            "validate_certs": true
        }
    },
    "msg": "Unable to upload.",
    "reason": [
        "HTTPSConnectionPool(host='esxi01', port=443): Max retries exceeded with url: /folder/VMware-Log-Insight-Agent-4.6.1-8587550.noarch.rpm?dsName=nfs_datastore (Caused by SSLError(SSLError(\"bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)\",),))",
        "HTTPSConnectionPool(host='esxi04', port=443): Max retries exceeded with url: /folder/VMware-Log-Insight-Agent-4.6.1-8587550.noarch.rpm?dsName=nfs_datastore (Caused by SSLError(SSLError(\"bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)\",),))",
        "HTTPSConnectionPool(host='esxi03', port=443): Max retries exceeded with url: /folder/VMware-Log-Insight-Agent-4.6.1-8587550.noarch.rpm?dsName=nfs_datastore (Caused by SSLError(SSLError(\"bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)\",),))",
        "HTTPSConnectionPool(host='esxi02', port=443): Max retries exceeded with url: /folder/VMware-Log-Insight-Agent-4.6.1-8587550.noarch.rpm?dsName=nfs_datastore (Caused by SSLError(SSLError(\"bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)\",),))"
    ]
}
```
I use custom certificates, but I don't get such errors when using other modules with ` "validate_certs": true`.

Rest LGTM

**Comment by Akasurde** [comment]
@ckotte Do you have ESXi's certificate installed in your jump host or delegate_host ?

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (ckotte) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: You should add real values: https://github.com/ansible/ansible/pull/47271#discussion_r226405450

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (ckotte) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py:50]: You should add the type for each option: https://github.com/ansible/ansible/pull/47271#discussion_r226381898.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (ckotte) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Would it be possible to add a verbose message? 
E.g. `"msg": File 'abc' already uploaded` or `"msg": File 'abc' uploaded`

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: All modules in VMware space uses Jinja variables like the current implementation. I would stick to this naming scheme as this will make all modules same and readable. Adding arbitary values does not make sense to me.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: +1 for this.

**Comment by ckotte** [comment]
> @ckotte Do you have ESXi's certificate installed in your jump host or delegate_host ?

The CA chain is stored in `/etc/pki/ca-trust/source/anchors/cachain.pem` in my CentOS 7 Docker image. Also, curl doesn't throw an error. 
```
# curl https://esxi01.fqdn.local
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">

<html lang="en">
<head>
    <meta http-equiv="content-type" content="text/html; charset=utf8">
    <meta http-equiv="refresh" content="0;URL='/ui'"/>
</head>
</html>
```

However, I need to use my local Docker image instead of the jump server I usually use (`delegate_to: "{{ jump_server }}"` instead of localhost). Is there a way to copy the files to the jump server and to the ESXi host in one step to test the jump server?

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: @dagwieers told me to use real values. It makes sense. Instead of `'{{ source_file_name }}'` and `'{{ destination_file_name }}'`, we should use `../folder/this-file.rpm` and `this-file.rpm`. Maybe you should take the opportunity and discuss https://github.com/ansible/ansible/issues/43433 in with the core team? I agree that we should have the same scheme across the modules.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (jeking3) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: I avoided `module_utils/urls.py` for a number of reasons - it appears that a convenience function called `fetch_url` exists to make it a little easier to issue a `GET`, but there is no corresponding `PUT` wrapper.  `requests` is used liberally throughout other modules like google's, and it's much easier to deal with.  Is your comment above a requirement for the vmware modules?

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (jeking3) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py:44]: Sure, someone should probably make a PR to standardize this across the modules that are there.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (jeking3) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py:50]: Agree, will do.

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (jeking3) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Yes, can do.

**Comment by jeking3** [comment]
When committing changes in response to comments do you prefer additional commits which you will eventually just squash, or do you want me to continue squash and force push so there is one commit?

**Comment by jeking3** [comment]
On the SSL issues, perhaps my use of `requests` is to blame here.

**Comment by ckotte** [comment]
> When committing changes in response to comments do you prefer additional commits which you will eventually just squash, or do you want me to continue squash and force push so there is one commit?

I only do `git commit -v --amend` and `git push --force` and no one complained yet.

**Comment by dagwieers** [comment]
> When committing changes in response to comments do you prefer additional commits which you will eventually just squash, or do you want me to continue squash and force push so there is one commit?

There's no consensus yet. For people reviewing, commenting and re-reviewing PRs (especially, complex PRs) it's probably wise to keep separate commits, so they can only review the changes compared to their comments. On the other hand, when I am preparing a PR and when I have sets of changes that are related, I often amend existing commits until I get it right. This helps to roll back to a previous state if need be. (Or show different possible implementations).

When a PR is being merged, the commits are usually squashed into one commit. So there's no need for the contributor to do that. You can save yourself the trouble.

**Comment by pdellaert** [comment]
> When committing changes in response to comments do you prefer additional commits which you will eventually just squash, or do you want me to continue squash and force push so there is one commit?

In my opinion, it is better to just keep adding commits, as @dagwieers explains, it helps in tracking changes from PR Reviews.

Once it is ready to be merged, the person merging it, can squash and merge it into a single commit.

**Comment by dagwieers** [comment]
So the long-term solution for copying files to vCenter or ESXi will be part of the vCenter/ESXi connection plugin. This would allow to use e.g. **copy** and **fetch** rather than dedicated modules.

**Comment by jeking3** [comment]
@dagwieers what does that mean for this PR and the enhanced functionality it provides?

**Comment by jeking3** [comment]
Switching from requests to the built-in urls.py, will submit an update shortly.  I think that's the last thing to do?  (Well, apart from tests, which I've only done manually...)

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (ckotte) [reply lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: You should at least add real values for `datastore`, `source`, and `destination`.

**Comment by ckotte** [comment]
> Switching from requests to the built-in urls.py, will submit an update shortly. I think that's the last thing to do? (Well, apart from tests, which I've only done manually...)

At least I have still certificate validation errors and a verbose message would be nice. You need to add tests for vcsim. Not sure if datastore upload is possible. You will find it our after you pushed the tests :)

**Comment by jeking3** [comment]
Resolved the failures reported above in the last push...

**Comment by jeking3** [comment]
So, what's the next step here?

**Comment by Im0** [comment]
Hi @jeking3 - any chance of an integration test?  Although, I'm wondering if datastore upload would work with vcsim?  The other integration tests maybe of use in putting something together: https://github.com/ansible/ansible/tree/devel/test/integration/targets

**Comment by jeking3** [comment]
@Im0 I'm big on testing.  I'll see what I can do.  I would prefer to see unit tests with MagicMock in the VMware modules, as the integration tests against a simulator are unlikely to exercise the failure handling paths well.  That's a big miss in the project overall.  I may not get around to this for a little while given my current deliverable schedule however.  I have been using this module quite a bit over the last few weeks and it's been reliable for my local uses.

**Comment by jeking3** [comment]
@Im0 does the integration test environment support mocking urllib responses?  This module creates URLs with ESXi names (DNS or IP) and contacts them directly through the built-in urls module_utils.  I think I would need to stand up some sort of https server.  I'm also not sure if vcsim support acquiring a service ticket.

**Comment by dagwieers** [comment]
@jeking3 I noticed in your initial summary you stated:

> The existing vsphere_copy does not directly copy to an available ESXi host...

But like many modules, you can specify an ESXi host as the host, and it should directly upload to that host using the same API your module is using. That was also the original reason to call this **vsphere_**  as it works both for vCenter as ESXi directly.

Also note that **vsphere_copy** was together with **vsphere_guest** one of the earliest modules created to manage VMware infrastructure. Using resp. the vsphere folder API and the pysphere library. (In the days when pyvmomi did not exist)

**Comment by dagwieers** [comment]
For integration tests, I make sure that those I write (lacking the real deal) can be run by anyone to verify their changes work in all conditions using `ansible-test`. Even if not integrated in CI, at least one can verify herself with no effort.

**Comment by jeking3** [comment]
I'll have to test the direct-to-ESXi Host part.  In my experience, ease of use comes from handling the cluster nature of vSphere automatically.  In this case, connect to vCenter and we'll try all of the hosts with that datastore mounted until one is responsive and handles the I/O.

**Comment by corbinstuard** [comment]
This looks good, what's holding it up?

**Comment by jeking3** [comment]
I had talked about adding integration tests for it but I have not done that.  I have been using this code for a while now and it seems fairly stable but I haven't tested all the corner cases, for example when an ESXi host is down in a cluster it is supposed to try another host.  The code is there but it may not have been exercised before.  I'm not convinced the simulator (vcsim) used in integration testing would be able to test this logic however.

@dagwieers would like folks to be able to point it at an ESXi Host directly (presumably so one could bootstrap installation of vCenter, or work in a vCenter-less environment).

File: lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py
Comment (pdellaert) [inline lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py]: Seems like a bug? `self.datacenter` doesn't exist
```suggestion
            datastore_name, datacenter=self._datacenter)
```

**Review by pdellaert — CHANGES_REQUESTED** [review]
Overall a nice addition, one bug (i think). And integration tests should be provided as a best practice.

I agree that an option to directly go to a ESXi server would be good to have, but i don't feel like that should stop this module from being added

**Comment by jeking3** [comment]
I suppose it might be possible to use an integration test due to some simulator support for acquiring tickets, but I don't know if the simulator will simulate an ESXi Host accepting a file push.

https://github.com/vmware/govmomi/search?q=AcquireGenericServiceTicket&unscoped_q=AcquireGenericServiceTicket

**Comment by ulysses-31** [comment]
It would be great in future if we could also support download/fetch from datastore. I have a use case where I build VM templates in one vSphere environment which I need to download for upload to other vSphere environments.

**Pull request closed without merging**
